youtubeBot/bot.py

```diff
@@ -34,6 +59,29 @@ def send_text(self, text):
 class YoutubeBot(Bot):
     pass  # TODO your code here!
 
+    # no need to define a constructor , default constructor will be generated .
+
+    # static variable -- Dictionary hold previous videos as pairs {message: video path}.
+    downloadedVideos = {}
+
+    def message_handler(self, message):
+
+        """Main messages handler"""
+        # step 1 - download the requested video.
+
+        if not YoutubeBot.downloadedVideos.__contains__(message):
+            print("Downloading Video ...\n")
+            video_name = download_youtube_file(message)[0]
+            YoutubeBot.downloadedVideos[message] = video_name
```

Comment:
If downloadVideos would be in __init__ as a class attribute, you could do `self.downloadVideos[message]`